chartutils/carvel/carvel.go

```diff
@@ -0,0 +1,168 @@
+type Website struct {
+	URL string `json:"url,omitempty"`
+}
+
+// Bundle Metadata
+const (
+	BundleAPIVersion = "imgpkg.carvel.dev/v1alpha1"
+	BundleKind       = "Bundle"
+)
+
+// BundleVersion with detailsa bout the Carvel bundle version
+type BundleVersion struct {
+	APIVersion string `json:"apiVersion"` // This generated yaml, but due to lib we need to use `json`
+	Kind       string `json:"kind"`       // This generated yaml, but due to lib we need to use `json`
+}
+
+// Metadata for a Carvel bundle
+type Metadata struct {
+	Version  BundleVersion
+	Metadata map[string]string `json:"metadata,omitempty"`
+	Authors  []Author          `json:"authors,omitempty"`
+	Websites []Website         `json:"websites,omitempty"`
+}
+
+// ToYAML serializes the Carvel bundle into YAML
+func (il *Metadata) ToYAML(w io.Writer) error {
+	enc := yaml.NewEncoder(w)
+	enc.SetIndent(2)
+
+	return enc.Encode(il)
+}
+
+// CarvelBundleFromYAMLFile Deserializes a string into a Metadata struct
+func CarvelBundleFromYAMLFile(file string) (*Metadata, error) {
+	fh, err := os.Open(file)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open Images.lock file: %v", err)
+	}
+	defer fh.Close()
+	return CarvelBundleFromYAML(fh)
+}
+
+// CarvelBundleFromYAML reads a Carvel metadata bundled from the YAML read from r
+func CarvelBundleFromYAML(r io.Reader) (*Metadata, error) {
+	metadata := &Metadata{
+		Version: BundleVersion{
+			APIVersion: BundleAPIVersion,
+			Kind:       BundleKind,
+		},
+		Metadata: map[string]string{},
+		Authors:  []Author{},
+		Websites: []Website{},
+	}
+	dec := yaml.NewDecoder(r)
+	if err := dec.Decode(metadata); err != nil {
+		return nil, fmt.Errorf("failed to load Carvel bundle: %v", err)
+	}
+
+	return metadata, nil
+}
+
+// NewCarvelBundle returns a new carvel bundle Metadata instance
+func NewCarvelBundle() *Metadata {
+
+	return &Metadata{
+		Version: BundleVersion{
+			APIVersion: BundleAPIVersion,
+			Kind:       BundleKind,
+		},
+		Metadata: map[string]string{},
+		Authors:  []Author{},
+		Websites: []Website{},
+	}
+}
+
+// PrepareBundleMetadata builds and sets a new Carvel bundle struct
+func PrepareBundleMetadata(chartPath string, lock *imagelock.ImagesLock) (*Metadata, error) {
+
+	bundleMetadata := NewCarvelBundle()
+
+	chart, err := chartutils.LoadChart(chartPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load chart: %w", err)
+	}
+
+	for _, maintainer := range chart.Metadata.Maintainers {
+		author := Author{
+			Name: maintainer.Name,
+		}
+		if maintainer.Email != "" {
+			author.Email = maintainer.Email
+		}
+		bundleMetadata.Authors = append(bundleMetadata.Authors, author)
+	}
+	for _, source := range chart.Metadata.Sources {
+		website := Website{
+			URL: source,
+		}
+		bundleMetadata.Websites = append(bundleMetadata.Websites, website)
+	}
+
+	bundleMetadata.Metadata["name"] = lock.Chart.Name
+	for key, value := range chart.Metadata.Annotations {
+		if key != "images" {
```

Comment:
I'm not 100% sure why we are not interested in the "images" key, but should this be the externally configured annotationsKey? Or at least use  imagelock.DefaultAnnotationsKey?

Reply:
It would be counter-intuitive for Carvel bundles to have the images showing up in the bundle as they are expected to show up in the `images.yml` file. For example, when removing that check, this is how a bundle would look like:

```yaml
bash-3.2$ cat examples/postgresql/.imgpkg/bundle.yml
version:
  apiversion: imgpkg.carvel.dev/v1alpha1
  kind: Bundle
metadata:
  category: Database
  images: |
    - name: os-shell
      image: docker.io/bitnami/os-shell:11-debian-11-r43
    - name: postgres-exporter
      image: docker.io/bitnami/postgres-exporter:0.13.2-debian-11-r25
    - name: postgresql
      image: docker.io/bitnami/postgresql:15.4.0-debian-11-r10
  licenses: Apache-2.0
  name: postgresql
authors:
  - name: VMware, Inc.
    email: ""
websites:
  - url: https://github.com/bitnami/charts/tree/main/bitnami/postgresql
```